syntax = "proto3";

package kmeans;

service KMeansService {
  // Master to Mapper communication
  rpc ProcessMap(TaskRequest) returns (TaskResponse) {}
  // Master to Reducer communication
  rpc ProcessReduce(TaskRequest) returns (TaskResponse) {}
  // Reducer to Mapper (to fetch intermediate data)
  rpc FetchData(FetchRequest) returns (FetchResponse) {}
}

message TaskRequest {
  repeated double centroids = 2;
  string dataChunk = 3;
  int32 numReducers = 4;
}

message TaskResponse {
  bool success = 1;
  string message = 2;
}

message FetchRequest {
  string key = 1;
}

message FetchResponse {
  repeated double dataPoints = 1;
}
